sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/CbsTokenProvider.cs

```diff
@@ -62,14 +83,94 @@ internal sealed class CbsTokenProvider : ICbsTokenProvider
         /// <param name="namespaceAddress">The address of the namespace to be authorized.</param>
         /// <param name="appliesTo">The resource to which the token should apply.</param>
         /// <param name="requiredClaims">The set of claims that are required for authorization.</param>
+        ///
         /// <returns>The token to use for authorization.</returns>
         ///
         public async Task<CbsToken> GetTokenAsync(Uri namespaceAddress,
                                                   string appliesTo,
-                                                  string[] requiredClaims)
+                                                  string[] requiredClaims) =>
+            Credential switch
+            {
+                _ when Credential.IsSharedAccessCredential => await AcquireSharedAccessTokenAsync().ConfigureAwait(false),
+                _ => await AcquireJwtTokenAsync().ConfigureAwait(false)
+            };
+
+        /// <summary>
+        ///   Performs the task needed to clean up resources used by the <see cref="CbsTokenProvider" />.
+        /// </summary>
+        ///
+        public void Dispose() => TokenAcquireSemaphore?.Dispose();
+
+        /// <summary>
+        ///   Acquires a token for a JWT-type credential.
+        /// </summary>
+        ///
+        /// <returns>The <see cref="CbsToken"/> to use for authorization.</returns>
+        ///
+        private async Task<CbsToken> AcquireJwtTokenAsync()
+        {
+            var token = Volatile.Read(ref _cachedJwtToken);
```

Comment:
AMQP authorization is refreshed only when a token is due to expire, a token was rejected, or a new connection/link is being established, most often taking place on a background thread.  Because this isn't on a performance-critical path called during every network operation, ensuring a longer-lived token is more valuable than quick token acquisition for overall throughput. 

Rather than doing something more complicated, like returning a short-lived token and requesting a fresh token in the background, I believe that we're better served by requesting the token on-demand and just synchronizing to ensure we don't make redundant requests.   As a result, this is intentionally simple, just using the good old lock+double-check pattern.